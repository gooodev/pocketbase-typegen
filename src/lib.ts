import {
  ALIAS_TYPE_DEFINITIONS,
  AUTH_SYSTEM_FIELDS_DEFINITION,
  BASE_SYSTEM_FIELDS_DEFINITION,
  DATE_STRING_TYPE_NAME,
  EXPAND_TYPE_COMMENT,
  EXPORT_COMMENT,
  RECORD_ID_STRING_NAME,
  RECORD_TYPE_COMMENT,
  RESPONSE_TYPE_COMMENT,
} from "./constants"
import { CollectionRecord, FieldSchema } from "./types"
import {
  fieldNameToGeneric,
  getGenericArgString,
  getGenericArgStringWithDefault,
} from "./generics"
import {
  getOptionEnumName,
  getOptionValues,
  sanitizeFieldName,
  toPascalCase,
} from "./utils"

const pbSchemaTypescriptMap = {
  text: "string",
  number: "number",
  bool: "boolean",
  email: "string",
  url: "string",
  date: DATE_STRING_TYPE_NAME,
  select: (fieldSchema: FieldSchema, collectionName: string) => {
    // pocketbase v0.8+ values are required
    const valueType = fieldSchema.options.values
      ? getOptionEnumName(collectionName, fieldSchema.name)
      : "string"
    return fieldSchema.options.maxSelect && fieldSchema.options.maxSelect > 1
      ? `${valueType}[]`
      : valueType
  },
  json: (fieldSchema: FieldSchema) =>
    `null | ${fieldNameToGeneric(fieldSchema.name)}`,
  file: (fieldSchema: FieldSchema) =>
    fieldSchema.options.maxSelect && fieldSchema.options.maxSelect > 1
      ? "string[]"
      : "string",
  relation: (fieldSchema: FieldSchema) =>
    fieldSchema.options.maxSelect && fieldSchema.options.maxSelect === 1
      ? RECORD_ID_STRING_NAME
      : `${RECORD_ID_STRING_NAME}[]`,
  // DEPRECATED: PocketBase v0.8 does not have a dedicated user relation
  user: (fieldSchema: FieldSchema) =>
    fieldSchema.options.maxSelect && fieldSchema.options.maxSelect > 1
      ? `${RECORD_ID_STRING_NAME}[]`
      : RECORD_ID_STRING_NAME,
}

export function generate(results: Array<CollectionRecord>) {
  const collectionNames: Array<string> = []
  const recordTypes: Array<string> = []
  const expandTypes: Array<string> = [EXPAND_TYPE_COMMENT]
  const responseTypes: Array<string> = [RESPONSE_TYPE_COMMENT]

  results
    .sort((a, b) => {
      if (a.name < b.name) {
        return -1
      }
      if (a.name > b.name) {
        return 1
      }
      return 0
    })
    .forEach((row) => {
      if (row.name) collectionNames.push(row.name)
      if (row.schema) {
        recordTypes.push(createRecordType(row.name, row.schema))
        responseTypes.push(createResponseType(row))
        expandTypes.push(createExpandType(results, row))
      }
    })
  const sortedCollectionNames = collectionNames

  const fileParts = [
    EXPORT_COMMENT,
    createCollectionEnum(sortedCollectionNames),
    ALIAS_TYPE_DEFINITIONS,
    BASE_SYSTEM_FIELDS_DEFINITION,
    AUTH_SYSTEM_FIELDS_DEFINITION,
    RECORD_TYPE_COMMENT,
    ...recordTypes,
    expandTypes.join("\n"),
    responseTypes.join("\n"),
    createCollectionRecords(sortedCollectionNames),
  ]

  return fileParts.join("\n\n")
}

export function createCollectionEnum(collectionNames: Array<string>) {
  const collections = collectionNames
    .map((name) => `\t${toPascalCase(name)} = "${name}",`)
    .join("\n")
  const typeString = `export enum Collections {
${collections}
}`
  return typeString
}

export function createCollectionRecords(collectionNames: Array<string>) {
  const nameRecordMap = collectionNames
    .map((name) => `\t${name}: ${toPascalCase(name)}Record`)
    .join("\n")
  return `export type CollectionRecords = {
${nameRecordMap}
}`
}

export function createRecordType(
  name: string,
  schema: Array<FieldSchema>
): string {
  const selectOptionEnums = createSelectOptions(name, schema)
  const typeName = toPascalCase(name)
  const genericArgs = getGenericArgStringWithDefault(schema)
  const fields = schema
    .map((fieldSchema: FieldSchema) => createTypeField(name, fieldSchema))
    .join("\n")

  return `${selectOptionEnums}export type ${typeName}Record${genericArgs} = {
${fields}
}`
}

export function createResponseType(collectionSchemaEntry: CollectionRecord) {
  const { name, schema, type } = collectionSchemaEntry
  const pascaleName = toPascalCase(name)
  const genericArgsWithDefaults = getGenericArgStringWithDefault(schema)
  const genericArgs = getGenericArgString(schema)
  return `export type ${pascaleName}Response${genericArgsWithDefaults} = ${pascaleName}Record${genericArgs} & BaseSystemFields<${pascaleName}ExpandType>${
    type === "auth" ? " & AuthSystemFields" : ""
  }`
}

export function createExpandType(
  allRecords: CollectionRecord[],
  targetRecord: CollectionRecord
) {
  const [directExpandTypes, directExpandConsts] = createDirectExpand(
    allRecords,
    targetRecord
  )
  const [indirectExpandTypes, indirectExpandConsts] = createIndirectExpand(
    allRecords,
    targetRecord
  )

  const expandTypes = [...directExpandTypes, ...indirectExpandTypes]
  const expandConsts = [...directExpandConsts, ...indirectExpandConsts]

  const hasExpandTypes = expandTypes.length > 0
  const pascaleName = toPascalCase(targetRecord.name)

  const expandTypesString = `type ${pascaleName}ExpandType = {
${hasExpandTypes ? expandTypes.join("\n") : "\t// Doesn't have any relation"}
}`

  const expandConstString =
    hasExpandTypes &&
    `export const ${pascaleName}ExpandKeys = {
  ${expandConsts.join(",\n")}
}`

  return [expandTypesString, expandConstString].filter(Boolean).join("\n")
}

export function createDirectExpand(
  allRecords: CollectionRecord[],
  targetRecord: CollectionRecord
) {
  const expandTypes: string[] = []
  const expandConsts: string[] = []

  targetRecord.schema
    .filter(({ type }) => type === "relation")
    .forEach(({ name, options }) => {
      const expandedRecord = allRecords.find(
        (record) => record.id === options?.collectionId
      )
      if (!expandedRecord) {
        return
      }
      const expandType =
        options.maxSelect && options.maxSelect > 1
          ? `${toPascalCase(expandedRecord.name)}Response[]`
          : `${toPascalCase(expandedRecord.name)}Response`
      expandTypes.push(`\t${name}: ${expandType}`)

      const expandConst = `\t${name}: "${name}"`
      expandConsts.push(expandConst)
    })
  return [expandTypes, expandConsts]
}

export function createIndirectExpand(
  allRecords: CollectionRecord[],
  targetRecord: CollectionRecord
) {
  const expandTypes: string[] = []
  const expandConsts: string[] = []

  allRecords.forEach((record) => {
    record.schema
      .filter((field) => field.type === "relation")
      .forEach((field) => {
        if (field.options?.collectionId !== targetRecord.id) {
          return
        }
        if (field.options.maxSelect && field.options.maxSelect > 1) {
          return
        }
        const expandKeyName = `${record.name}(${field.name})`
        expandTypes.push(
          `\t"${expandKeyName}": ${toPascalCase(record.name)}Response[]`
        )
        expandConsts.push(`\t${record.name}: "${expandKeyName}"`)
      })
  })
  return [expandTypes, expandConsts]
}

export function createTypeField(
  collectionName: string,
  fieldSchema: FieldSchema
) {
  if (!(fieldSchema.type in pbSchemaTypescriptMap)) {
    throw new Error(`unknown type ${fieldSchema.type} found in schema`)
  }
  const typeStringOrFunc =
    pbSchemaTypescriptMap[
      fieldSchema.type as keyof typeof pbSchemaTypescriptMap
    ]

  const typeString =
    typeof typeStringOrFunc === "function"
      ? typeStringOrFunc(fieldSchema, collectionName)
      : typeStringOrFunc

  const fieldName = sanitizeFieldName(fieldSchema.name)
  const required = fieldSchema.required ? "" : "?"

  return `\t${fieldName}${required}: ${typeString}`
}

export function createSelectOptions(
  recordName: string,
  schema: Array<FieldSchema>
) {
  const selectFields = schema.filter((field) => field.type === "select")
  const typestring = selectFields
    .map(
      (field) => `export enum ${getOptionEnumName(recordName, field.name)} {
${getOptionValues(field)
  .map((val) => `\t"${val}" = "${val}",`)
  .join("\n")}
}\n`
    )
    .join("\n")
  return typestring
}
