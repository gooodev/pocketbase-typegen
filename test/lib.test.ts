import { CollectionRecord, FieldSchema } from "../src/types"
import {
  createCollectionEnum,
  createCollectionRecords,
  createDirectExpand,
  createExpandType,
  createIndirectExpand,
  createRecordType,
  createResponseType,
  createSelectOptions,
  createTypeField,
  generate,
} from "../src/lib"

const defaultFieldSchema: FieldSchema = {
  id: "abc",
  system: false,
  unique: false,
  options: {},
  name: "defaultName",
  required: true,
  type: "text",
}

describe("generate", () => {
  it("generates correct output given db input", () => {
    const collections: Array<CollectionRecord> = [
      {
        name: "books",
        id: "123",
        type: "base",
        system: false,
        listRule: null,
        viewRule: null,
        createRule: null,
        updateRule: null,
        deleteRule: null,
        schema: [
          {
            name: "title",
            type: "text",
            required: false,
            id: "xyz",
            system: false,
            unique: false,
            options: {},
          },
        ],
      },
    ]
    const result = generate(collections)
    expect(result).toMatchSnapshot()
  })
})

describe("createCollectionEnum", () => {
  it("creates enum of collection names", () => {
    const names = ["book", "magazine"]
    expect(createCollectionEnum(names)).toMatchSnapshot()
  })
})

describe("createCollectionRecord", () => {
  it("creates mapping of collection name to record type", () => {
    const names = ["book", "magazine"]
    expect(createCollectionRecords(names)).toMatchSnapshot()
  })
})

describe("createRecordType", () => {
  it("creates type definition for a record", () => {
    const name = "books"
    const schema: FieldSchema[] = [
      {
        system: false,
        id: "hhnwjkke",
        name: "title",
        type: "text",
        required: false,
        unique: false,
        options: { min: null, max: null, pattern: "" },
      },
    ]
    const result = createRecordType(name, schema)
    expect(result).toMatchSnapshot()
  })

  it("handles file fields with multiple files", () => {
    const name = "books"
    const schema: FieldSchema[] = [
      {
        system: false,
        id: "hhnwjkke",
        name: "avatars",
        type: "file",
        required: false,
        unique: false,
        options: { maxSelect: 2 },
      },
    ]
    const result = createRecordType(name, schema)
    expect(result).toMatchSnapshot()
  })
})

describe("createResponseType", () => {
  it("creates type definition for a response", () => {
    const row: CollectionRecord = {
      type: "base",
      id: "123",
      system: false,
      listRule: null,
      viewRule: null,
      createRule: null,
      updateRule: null,
      deleteRule: null,
      name: "books",
      schema: [
        {
          system: false,
          id: "hhnwjkke",
          name: "title",
          type: "text",
          required: false,
          unique: false,
          options: { min: null, max: null, pattern: "" },
        },
      ],
    }

    const result = createResponseType(row)
    expect(result).toMatchSnapshot()
  })

  it("handles file fields with multiple files", () => {
    const name = "books"
    const schema: FieldSchema[] = [
      {
        system: false,
        id: "hhnwjkke",
        name: "avatars",
        type: "file",
        required: false,
        unique: false,
        options: { maxSelect: 2 },
      },
    ]
    const result = createRecordType(name, schema)
    expect(result).toMatchSnapshot()
  })
})

describe("createTypeField", () => {
  it("handles required and optional fields", () => {
    expect(
      createTypeField("test_collection", {
        ...defaultFieldSchema,
        required: false,
      })
    ).toEqual("\tdefaultName?: string")
    expect(
      createTypeField("test_collection", {
        ...defaultFieldSchema,
        required: true,
      })
    ).toEqual("\tdefaultName: string")
  })

  it("converts default types to typescript", () => {
    expect(
      createTypeField("test_collection", {
        ...defaultFieldSchema,
      })
    ).toEqual("\tdefaultName: string")
    expect(
      createTypeField("test_collection", {
        ...defaultFieldSchema,
        name: "textField",
      })
    ).toEqual("\ttextField: string")
  })

  it("converts number type", () => {
    expect(
      createTypeField("test_collection", {
        ...defaultFieldSchema,
        name: "numberField",
        type: "number",
      })
    ).toEqual("\tnumberField: number")
  })

  it("converts bool type", () => {
    expect(
      createTypeField("test_collection", {
        ...defaultFieldSchema,
        name: "boolField",
        type: "bool",
      })
    ).toEqual("\tboolField: boolean")
  })

  it("converts email type", () => {
    expect(
      createTypeField("test_collection", {
        ...defaultFieldSchema,
        name: "emailField",
        type: "email",
      })
    ).toEqual("\temailField: string")
  })

  it("converts url type", () => {
    expect(
      createTypeField("test_collection", {
        ...defaultFieldSchema,
        name: "urlField",
        type: "url",
      })
    ).toEqual("\turlField: string")
  })

  it("converts date type", () => {
    expect(
      createTypeField("test_collection", {
        ...defaultFieldSchema,
        name: "dateField",
        type: "date",
      })
    ).toEqual("\tdateField: IsoDateString")
  })

  it("converts select type", () => {
    expect(
      createTypeField("test_collection", {
        ...defaultFieldSchema,
        name: "selectField",
        type: "select",
      })
    ).toEqual("\tselectField: string")
  })

  it("converts select type with value", () => {
    expect(
      createTypeField("test_collection", {
        ...defaultFieldSchema,
        name: "selectFieldWithOpts",
        type: "select",
        options: {
          values: ["one", "two", "three"],
        },
      })
    ).toEqual(`\tselectFieldWithOpts: TestCollectionSelectFieldWithOptsOptions`)
  })

  it("converts multi-select type", () => {
    expect(
      createTypeField("test_collection", {
        ...defaultFieldSchema,
        name: "selectField",
        type: "select",
        options: {
          maxSelect: 2,
        },
      })
    ).toEqual("\tselectField: string[]")
  })

  it("converts multi-select type with values", () => {
    expect(
      createTypeField("test_collection", {
        ...defaultFieldSchema,
        name: "selectFieldWithOpts",
        type: "select",
        options: {
          values: ["one", "two", "three"],
          maxSelect: 2,
        },
      })
    ).toEqual(
      `\tselectFieldWithOpts: TestCollectionSelectFieldWithOptsOptions[]`
    )
  })

  it("converts json type", () => {
    expect(
      createTypeField("test_collection", {
        ...defaultFieldSchema,
        name: "jsonField",
        type: "json",
      })
    ).toEqual("\tjsonField: null | TjsonField")
  })

  it("converts file type", () => {
    expect(
      createTypeField("test_collection", {
        ...defaultFieldSchema,
        name: "fileField",
        type: "file",
      })
    ).toEqual("\tfileField: string")
  })

  it("converts file type with multiple files", () => {
    expect(
      createTypeField("test_collection", {
        ...defaultFieldSchema,
        name: "fileField",
        type: "file",
        options: {
          maxSelect: 3,
        },
      })
    ).toEqual("\tfileField: string[]")
  })

  it("converts relation type", () => {
    expect(
      createTypeField("test_collection", {
        ...defaultFieldSchema,
        name: "relationField",
        type: "relation",
      })
    ).toEqual("\trelationField: RecordIdString[]")
  })

  it("converts relation type with multiple options", () => {
    expect(
      createTypeField("test_collection", {
        ...defaultFieldSchema,
        name: "relationFieldMany",
        type: "relation",
        options: {
          maxSelect: 3,
        },
      })
    ).toEqual("\trelationFieldMany: RecordIdString[]")
  })

  it("converts relation type with unset maxSelect", () => {
    expect(
      createTypeField("test_collection", {
        ...defaultFieldSchema,
        name: "relationFieldMany",
        type: "relation",
        options: {
          maxSelect: null,
        },
      })
    ).toEqual("\trelationFieldMany: RecordIdString[]")
  })

  // DEPRECATED: This was removed in PocketBase v0.8
  it("converts user relation type", () => {
    expect(
      createTypeField("test_collection", {
        ...defaultFieldSchema,
        name: "userRelationField",
        type: "user",
      })
    ).toEqual("\tuserRelationField: RecordIdString")
  })

  it("throws for unexpected types", () => {
    expect(() =>
      createTypeField("test_collection", {
        ...defaultFieldSchema,
        // @ts-ignore
        type: "unknowntype",
      })
    ).toThrowError("unknown type unknowntype found in schema")
  })
})

describe("createSelectOptions", () => {
  it("creates enum types for select options", () => {
    const name = "choose"
    const schema: FieldSchema[] = [
      {
        system: false,
        id: "hhnwjkke",
        name: "title",
        type: "select",
        required: false,
        unique: false,
        options: { values: ["one", "one", "two", "space space", "$@#*(&#%"] },
      },
    ]
    const result = createSelectOptions(name, schema)
    expect(result).toMatchSnapshot()
  })
})

describe("createDirectExpand", () => {
  const collections: Array<CollectionRecord> = [
    {
      name: "users",
      id: "123",
      type: "auth",
      system: false,
      listRule: null,
      viewRule: null,
      createRule: null,
      updateRule: null,
      deleteRule: null,
      schema: [
        {
          name: "name",
          type: "text",
          required: false,
          id: "xyz",
          system: false,
          unique: false,
          options: {},
        },
      ],
    },
    {
      name: "authors",
      id: "456",
      type: "base",
      system: false,
      listRule: null,
      viewRule: null,
      createRule: null,
      updateRule: null,
      deleteRule: null,
      schema: [
        {
          name: "user",
          type: "relation",
          required: false,
          id: "xyz",
          system: false,
          unique: false,
          options: {
            collectionId: "123",
            cascadeDelete: false,
            maxSelect: 1,
          },
        },
      ],
    },
    {
      name: "books",
      id: "789",
      type: "base",
      system: false,
      listRule: null,
      viewRule: null,
      createRule: null,
      updateRule: null,
      deleteRule: null,
      schema: [
        {
          name: "title",
          type: "text",
          required: false,
          id: "xyz",
          system: false,
          unique: false,
          options: {},
        },
        {
          name: "authors",
          type: "relation",
          required: false,
          id: "xyz",
          system: false,
          unique: false,
          options: {
            collectionId: "456",
            cascadeDelete: false,
            maxSelect: 2,
          },
        },
      ],
    },
    {
      name: "publishers",
      id: "abc",
      type: "base",
      system: false,
      listRule: null,
      viewRule: null,
      createRule: null,
      updateRule: null,
      deleteRule: null,
      schema: [
        {
          name: "company",
          type: "relation",
          required: false,
          id: "xyz",
          system: false,
          unique: false,
          options: {
            collectionId: "xxx",
            cascadeDelete: false,
            maxSelect: 1,
          },
        },
      ],
    },
  ]

  describe("createDirectExpand", () => {
    it("create expand type for direct relation", () => {
      const authorsCollection = collections[1]
      const result = createDirectExpand(collections, authorsCollection)
      expect(result).toMatchSnapshot()
    })

    it("create expand type for direct relation with two or more maxSelect", () => {
      const booksCollection = collections[2]
      const result = createDirectExpand(collections, booksCollection)
      expect(result).toMatchSnapshot()
    })

    it("not create expand type for unexist direct relation", () => {
      const publishersCollection = collections[3]
      const result = createDirectExpand(collections, publishersCollection)
      expect(result).toMatchSnapshot()
    })
  })

  describe("createIndirectExpand", () => {
    it("create expand type for indirect relation", () => {
      const usersCollection = collections[0]
      const result = createIndirectExpand(collections, usersCollection)
      expect(result).toMatchSnapshot()
    })

    it("not create expand type for indirect relation with two or more maxSelect", () => {
      const authorsCollection = collections[1]
      const result = createIndirectExpand(collections, authorsCollection)
      expect(result).toMatchSnapshot()
    })
  })

  describe("createExpand", () => {
    it("create expand type", () => {
      const result = createExpandType(collections, collections[0])
      expect(result).toMatchSnapshot()
    })
  })
})
