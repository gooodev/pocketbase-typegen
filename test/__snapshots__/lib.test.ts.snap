// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCollectionEnum creates enum of collection names 1`] = `
"export enum Collections {
	Book = "book",
	Magazine = "magazine",
}"
`;

exports[`createCollectionRecord creates mapping of collection name to record type 1`] = `
"export type CollectionRecords = {
	book: BookRecord
	magazine: MagazineRecord
}"
`;

exports[`createDirectExpand createDirectExpand create expand type for direct relation 1`] = `
[
  [
    "	user: UsersResponse",
  ],
  [
    "	user: "user"",
  ],
]
`;

exports[`createDirectExpand createDirectExpand create expand type for direct relation with two or more maxSelect 1`] = `
[
  [
    "	authors: AuthorsResponse[]",
  ],
  [
    "	authors: "authors"",
  ],
]
`;

exports[`createDirectExpand createDirectExpand not create expand type for unexist direct relation 1`] = `
[
  [],
  [],
]
`;

exports[`createDirectExpand createExpand create expand type 1`] = `
"type UsersExpandType = {
	"authors(user)": AuthorsResponse[]
}
export const UsersExpandKeys = {
  	authors: "authors(user)"
}"
`;

exports[`createDirectExpand createIndirectExpand create expand type for indirect relation 1`] = `
[
  [
    "	"authors(user)": AuthorsResponse[]",
  ],
  [
    "	authors: "authors(user)"",
  ],
]
`;

exports[`createDirectExpand createIndirectExpand not create expand type for indirect relation with two or more maxSelect 1`] = `
[
  [],
  [],
]
`;

exports[`createRecordType creates type definition for a record 1`] = `
"export type BooksRecord = {
	title?: string
}"
`;

exports[`createRecordType handles file fields with multiple files 1`] = `
"export type BooksRecord = {
	avatars?: string[]
}"
`;

exports[`createResponseType creates type definition for a response 1`] = `"export type BooksResponse = BooksRecord & BaseSystemFields<BooksExpandType>"`;

exports[`createResponseType handles file fields with multiple files 1`] = `
"export type BooksRecord = {
	avatars?: string[]
}"
`;

exports[`createSelectOptions creates enum types for select options 1`] = `
"export enum ChooseTitleOptions {
	"one" = "one",
	"two" = "two",
	"space space" = "space space",
	"$@#*(&#%" = "$@#*(&#%",
}
"
`;

exports[`generate generates correct output given db input 1`] = `
"/**
* This file was @generated using pocketbase-typegen
*/

export enum Collections {
	Books = "books",
}

// Alias types for improved usability
export type IsoDateString = string
export type RecordIdString = string

// System fields
export type BaseSystemFields<T> = {
	id: RecordIdString
	created: IsoDateString
	updated: IsoDateString
	collectionId: string
	collectionName: Collections
	expand?: T extends object ? T : never
}

export type AuthSystemFields = {
	email: string
	emailVisibility: boolean
	username: string
	verified: boolean
}

// Record types for each collection

export type BooksRecord = {
	title?: string
}

// Expand types are used to define the expand object in the PocketBase API
type BooksExpandType = {
	// Doesn't have any relation
}

// Response types include system fields and match responses from the PocketBase API
export type BooksResponse = BooksRecord & BaseSystemFields<BooksExpandType>

export type CollectionRecords = {
	books: BooksRecord
}"
`;
